src/server/input/default_configuration.cpp

```diff
@@ -326,9 +327,16 @@ std::shared_ptr<mi::DefaultInputDeviceHub> mir::DefaultServerConfiguration::the_
 std::shared_ptr<mi::KeyMapper> mir::DefaultServerConfiguration::the_key_mapper()
 {
     return key_mapper(
-       []()
+       [this]()
        {
-           return std::make_shared<mi::receiver::XKBMapper>();
+           return std::make_shared<mi::receiver::XKBMapper>([this](auto device_id, auto leds)
+           {
+               the_default_input_device_hub()->for_each_mutable_input_device([&](mir::input::Device& device)
+               {
+                    if (device.id() == device_id)
+                        device.set_leds(leds);
+               });
+           });
```

Comment:
We should probably capture `the_default_input_device_hub()`, not call it every time the mapping state changes.

And, surely, there could be a more direct way from `XKBMapper` to the corresponding device than scanning everything registered on the device hub? (E.g. a map `MirInputDeviceId` -> callback)

Reply:
The problem is that ``the_default_input_device_hub()` also relies on the `the_key_mapper()`, so it's a circular dependency. I will see if there is some other way around this, perhaps with the "registration" idea that you said either.